### PR TITLE
18 configure database

### PR DESCRIPTION
**Pull request by wojciechkapko**
Added EntityFramework (ORM)
Initial ORM configuration

Added Seeder class to seed data from our source file

### CODE REVIEW

File: DrinkManagerWeb/Data/Seeder.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+                .GetRequiredService<IServiceScopeFactory>().CreateScope();
+            var context = serviceScope
+                .ServiceProvider.GetService<DrinkAppContext>();
+
+            // Check if we can connect to the database
+            if (!context.Database.CanConnect())
+            {
+                // Create the database
+                context.Database.Migrate();
+                Console.WriteLine("Database does not exist...Creating");
```
Comment (PatrykSzwer) [inline DrinkManagerWeb/Data/Seeder.cs]: More logically it would be to replace those 2 lines with each other.

File: DrinkManagerWeb/Data/DrinkAppContext.cs [hunk 2]
```diff
@@ -1,15 +1,16 @@
...
-using System.Collections.Generic;
+using Microsoft.EntityFrameworkCore;
 
 namespace DrinkManagerWeb.Data
 {
-    public class DrinkAppContext
+    public class DrinkAppContext : DbContext
     {
-        public DrinkAppContext(DrinkLoader loader)
+        public DrinkAppContext(DbContextOptions<DrinkAppContext> options) : base(options)
```
Comment (PatrykSzwer) [inline DrinkManagerWeb/Data/DrinkAppContext.cs]: This should be in the BLL project, not in the web.

File: BLL/Drink.cs [hunk 3]
```diff
@@ -1,11 +1,18 @@
...
 
 namespace BLL
 {
     public class Drink
     {
+        [Key]
+        [DatabaseGenerated(DatabaseGeneratedOption.None)]
         public string Id { get; set; }
+
+        [Required]
```
Comment (PatrykSzwer) [inline BLL/Drink.cs:13]: I would think about more restrictions over this entity. I think it will not only be the Name that is required, and you can think about more general restrictions, like min-max length, etc.

**Review by PatrykSzwer — COMMENTED** [review]
From the configuration part, it looks fine, the next part is to create a bit of access layer to reduce amount of repeated code, and help others use the DB context.
I would like you to read this article: https://www.thereformedprogrammer.net/is-the-repository-pattern-useful-with-entity-framework-core/ 
and specifically, focus on the part, where DDD-styled entity classes are presented. There is an example of the Book entity with `AddReview` method described inside of it. I think that would be a cool feature of your app, to expose such functions for base operations on `Drink` to make use of the database layer much more friendly. Let me know if you have any questions and when you'll be done with this part, and I'll check again :) 

One thing, I know its not a part of this PR, but could you fix property in the Drink class called "isReviewed"?
Right now:
```
        public bool isReviewed
        {
            get
            {
                if (DrinkReview != null)
                {
                    return true;
                }
                return false;
            }
        }
```

Should be:
`public bool IsReviewed => DrinkReview != null;`

Hint: use IntelliSense function for it, (ctrl + dot, `Rename to` function) so it will automatically rename it everywhere.

**Review by PatrykSzwer — CHANGES_REQUESTED** [review]
Implementation of repository itself looks good, but I didn't test it so, second reviewer, please test it out. Also, please fix conflicts.

File: BLL/Data/Repositories/IDrinkRepository.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace BLL.Data.Repositories
+{
+    public interface IDrinkRepository
+    {
+        IQueryable<Drink> GetAllDrinks();
+
+        Task<Drink> GetDrinkById(string id);
+
+        public Task<Drink> FindDrink(Expression<Func<Drink, bool>> predicate);
```
Comment (PatrykSzwer) [inline BLL/Data/Repositories/IDrinkRepository.cs]: We're not using access modifiers in the interface definition!

File: BLL/Data/DrinkAppContext.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+namespace BLL.Data
+{
+    public class DrinkAppContext : DbContext
+    {
+        public DrinkAppContext(DbContextOptions<DrinkAppContext> options) : base(options)
+        {
+        }
+
+        public DbSet<Drink> Drinks { get; set; }
+        public DbSet<Ingredient> Ingredients { get; set; }
```
Comment (PatrykSzwer) [inline BLL/Data/DrinkAppContext.cs:12]: Ingredients are not database entities, only part of Drink entity. Remove this.

File: BLL/DrinkLoader.cs [hunk 6]
```diff
@@ -9,26 +10,30 @@ namespace BLL
...
-            }
-        }
+        // public void AddDrinksFromFile(List<Drink> currentDrinks, string path)
+        // {
+        //     var newDrinks = LoadFromFile(path);
+        //     if (newDrinks != null)
+        //     {
+        //         currentDrinks.AddRange(newDrinks);
+        //     }
+        // }
```
Comment (PatrykSzwer) [inline BLL/DrinkLoader.cs]: Commented code, please remove.

File: BLL/DrinkLoader.cs [hunk 7]
```diff
@@ -9,26 +10,30 @@ namespace BLL
...
 
-        private List<Drink> LoadFromFile(string path)
+        private FileLoaderDto LoadFromFile(string path)
         {
             var newDrinks = new List<Drink>();
+            var newIngredients = new List<Ingredient>();
+
+
+
```
Comment (PatrykSzwer) [inline BLL/DrinkLoader.cs]: Redundant empty lines.

File: BLL/DrinkLoader.cs [hunk 8]
```diff
@@ -37,107 +42,132 @@ private List<Drink> LoadFromFile(string path)
...
+                    newDrinks.Add(newDrink);
                 }
             }
             catch (Exception)
             {
                 throw new FileNotFoundException("File not found, maybe the path was incorrect?");
             }
 
-            return newDrinks;
+
```
Comment (PatrykSzwer) [inline BLL/DrinkLoader.cs]: Redundant empty line.

File: BLL/DrinkLoader.cs [hunk 9]
```diff
@@ -9,26 +10,30 @@ namespace BLL
 {
     public class DrinkLoader
     {
-        public List<Drink> InitializeDrinksFromFile()
+        public FileLoaderDto InitializeDrinksFromFile()
```
Comment (PatrykSzwer) [inline BLL/DrinkLoader.cs]: Please revisit this again and make sure that entire solution is building correctly without errors. As we talked probably there is no point now in returning DTO by this method, since ingredients are part of the drink entity now.

File: BLL/PaginatedList.cs [hunk 10]
```diff
@@ -44,10 +46,10 @@ public string IsOnCurrentPage(int pageIndex, int currentPageIndex)
...
 
-        public static PaginatedList<T> CreatePaginatedList(IQueryable<T> source, int pageIndex, int pageSize)
+        public static async Task<PaginatedList<T>> CreatePaginatedList(IQueryable<T> source, int pageIndex, int pageSize)
         {
-            var count = source.Count();
-            var items = source.Skip((pageIndex - 1) * pageSize).Take(pageSize).ToList();
+            var count = await source.CountAsync();
+            var items = await source.Skip((pageIndex - 1) * pageSize).Take(pageSize).ToListAsync();
             return new PaginatedList<T>(items, count, pageIndex, pageSize);
         }
```
Comment (PatrykSzwer) [inline BLL/PaginatedList.cs]: Make sure that those changes are required, they don't seem to be part of this PR subject.

File: BLL/SearchDrink.cs [hunk 11]
```diff
@@ -1,102 +1,97 @@
...
+        // {
+        //     foreach (Drink drink in drinks)
+        //     {
+        //         if (drink.AlcoholicInfo == alcoholicInfo)
+        //         {
+        //             contemporaryList.Add(drink);
+        //         }
+        //     }
+        //     return contemporaryList;
+        // }
```
Comment (PatrykSzwer) [inline BLL/SearchDrink.cs]: This code should be uncommented to make Console app build'able again.

**Review by PatrykSzwer — CHANGES_REQUESTED** [review]
Check comments.

Please make sure that the entire solution is building before the next review request.

**Review by PatrykSzwer — APPROVED** [review]
LGTM, when someone else confirms that it works for them, you can merge!

**Review by marcin-sieminski — APPROVED** [review]
